src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -0,0 +1,264 @@
+    path_prefix: String,
+    /// The desired file size. A new file upload will be started
+    /// when the size exceeds this amount.
+    max_file_size: u64,
+    /// The aws sdk config.
+    sdk_config: SdkConfig,
+    /// Multi-part uploader for the current file.
+    /// Keeping the uploader in an `Option` to later take owned value.
+    current_file_uploader: Option<S3MultiPartUploader>,
+    /// Temporary buffer to store the encoded bytes.
+    buf: Vec<u8>,
+}
+
+impl CopyToS3Uploader {
+    fn new(
+        sdk_config: SdkConfig,
+        connection_details: CopyToS3Details,
+        file_name_prefix: String,
+    ) -> CopyToS3Uploader {
+        let (bucket, path_prefix) = Self::extract_s3_bucket_path(&connection_details.prefix);
+        CopyToS3Uploader {
+            desc: connection_details.desc,
+            sdk_config,
+            format: connection_details.format,
+            file_name_prefix,
+            bucket,
+            path_prefix,
+            max_file_size: connection_details.max_file_size,
+            file_index: 0,
+            current_file_uploader: None,
+            buf: Vec::new(),
+        }
+    }
+
+    /// Creates the uploader for the next file which starts the multi part upload.
+    async fn start_new_file(&mut self) -> Result<(), anyhow::Error> {
+        self.flush().await?;
+        assert!(self.current_file_uploader.is_none());
+
+        self.file_index += 1;
+        // TODO: remove hard-coded file extension .csv
+        let file_path = format!(
+            "{}/{}-{:04}.csv",
+            self.path_prefix, self.file_name_prefix, self.file_index
+        );
+
+        let bucket = self.bucket.clone();
+        info!(
+            "starting upload at bucket: {}, file {}",
+            &bucket, &file_path
+        );
+        let uploader = S3MultiPartUploader::try_new(
+            &self.sdk_config,
+            bucket,
+            file_path,
+            S3MultiPartUploaderConfig {
+                part_size_limit: ByteSize::mib(10).as_u64(),
+                file_size_limit: self.max_file_size,
+            },
+        )
+        .await?;
+        self.current_file_uploader = Some(uploader);
+        Ok(())
+    }
+
+    fn extract_s3_bucket_path(prefix: &str) -> (String, String) {
+        let uri = Uri::from_str(prefix).expect("valid s3 url");
+        let bucket = uri.host().expect("s3 bucket");
+        let path = uri.path().trim_start_matches('/').trim_end_matches('/');
+        (bucket.to_string(), path.to_string())
+    }
+
+    /// Finishes any remaining in-progress upload.
+    async fn flush(&mut self) -> Result<(), anyhow::Error> {
+        if let Some(uploader) = self.current_file_uploader.take() {
+            uploader.finish().await?;
+        }
+        Ok(())
+    }
+
+    /// Appends the row to the in-progress upload or creates a new upload if it
+    /// will exceed the max file size.
+    async fn append_row(&mut self, row: &Row) -> Result<(), anyhow::Error> {
```

Comment:
To maintain performance, it might be a good idea to write this in terms of batches of rows to upload, instead of uploading potentially individual rows.

Reply:
Actually it will not be uploading individual rows. The `S3MultiPartUploader.add_chunk` does it's own buffering and will upload a single part once it hits a threshold. I will clarify it further in the comments.

Reply:
Expanded the comment here.